File: Client/src/types/Product.ts
import { type Category } from "./Category";
import { type Brand } from "./Brand";
import { type ProductCommentResponse } from "./ProductComment";
import { type ProductImage } from "./ProductImage";
import { type Specification} from "./Specification";
import type { MessageResponse } from "./MessageReponse";
import type { ProductColor } from "./ProductColor";
import type { ProductModel } from "./ProductModel";

export interface Product {
    productId: number;
    name: string;
    slug: string;
    warranty: string;
    statusProduct: boolean;
    price: number;
    discount: number;
    priceAfterDiscount: number;
    endDateDiscount: string | null;
    views: number;
    dateOfManufacture: string | null;
    madeIn: string | null;
    promotionalGift?: string | null;
    photo: string;
    description: string;
    category: Category;
    brand: Brand;
    productComments: ProductCommentResponse[];
    productColors: ProductColor[];
    productModels: ProductModel[];
    productImages: ProductImage[];
    specifications: Specification[];
    relatedProducts: Product[];
}

export interface SearchProductProps extends MessageResponse{
    products?: Product[];
    initialSort?: string;
}

export interface ProductForm {
    id: string | number;
    name?: string;
    slug?: string;
    warranty?: string;
    statusProduct?: string;
    initialCost?: number;
    price?: number;
    discount?: number;
    priceAfterDiscount: number;
    endDateDiscount?: string | null;
    dateOfManufacture?: string | null;
    madeIn?: string | null;
    views?: number;
    promotionalGift?: string | null;
    description?: string;
    photo?: string;
    photoUpload?: File;
    categoryId?: number | string;
    brandId?: number | string;
    createDate?: string;
    createdBy?: string;
    updateDate?: string;
    updatedBy?: string;
}

export interface ProductFormProp {
    productInfor: ProductForm;
    productColors?: ProductColor[];
    specifications?: Specification[];
    productImages?: ProductImage[];
    productModels?: ProductModel[];
}

File: Domain/Entities/Product.cs
using DTech.Domain.Entities;
using DTech.Domain.Enums;
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace DTech.Domain.Entities;

public partial class Product
{
    [Key]
    [Display(Name = "ID")]
    [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
    public int ProductId { get; set; }

    [Display(Name = "Brand")]
    [Required(ErrorMessage = "Please select brand of product")]
    public int? BrandId { get; set; }

    [Display(Name = "Category")]
    [Required(ErrorMessage = "Please select category of product")]
    public int? CategoryId { get; set; }

    [Required(ErrorMessage = "Please enter name of product")]
    public string? Name { get; set; }

    public string? Slug { get; set; }

    [Required(ErrorMessage = "Please enter warranty of product")]
    public string? Warranty { get; set; }

    [Display(Name = "Product Status")]
    public bool? StatusProduct { get; set; }

    [Required(ErrorMessage = "Please enter initial cost of product")]
    [Display(Name = "Initial Cost")]
    public decimal? InitialCost { get; set; }

    [Required(ErrorMessage = "Please enter price of product")]
    public decimal? Price { get; set; }

    public decimal? Discount { get; set; }

    public decimal? PriceAfterDiscount { get; set; }

    [Display(Name = "Discount End Date")]
    public DateOnly? EndDateDiscount { get; set; }

    public int? Views { get; set; }

    [Display(Name = "Date of Manufacture")]
    public DateOnly? DateOfManufacture { get; set; }

    [Display(Name = "Made In")]
    public string? MadeIn { get; set; }

    [Display(Name = "Promotion Gift")]
    public string? PromotionalGift { get; set; }

    public string? Photo { get; set; }

    public string? Description { get; set; }

    [Display(Name = "Update Date")]
    public DateTime? UpdateDate { get; set; }

    [Display(Name = "Created By")]
    public string? CreatedBy { get; set; }

    [Display(Name = "Create Date")]
    public DateTime? CreateDate { get; set; }

    [Display(Name = "Updated By")]
    public string? UpdatedBy { get; set; }

    public StatusEnums Status { get; set; } = StatusEnums.Available;

    public virtual Brand? Brand { get; set; }

    public virtual Category? Category { get; set; }

    public virtual ICollection<OrderProduct> OrderProducts { get; set; } = [];

    [Display(Name = "Comment")]
    public virtual ICollection<ProductComment> ProductComments { get; set; } = [];

    public virtual ICollection<ProductImage> ProductImages { get; set; } = [];

    public virtual ICollection<Specification> Specifications { get; set; } = [];
}
